src/encoding.rs

```diff
@@ -2,6 +2,502 @@
+    fn from(e: text::LabelEncoder<'a>) -> Self {
+        Self::Text(e)
+    }
+}
+
+impl<'a> From<proto::LabelEncoder<'a>> for LabelEncoder<'a> {
+    fn from(e: proto::LabelEncoder<'a>) -> Self {
+        Self::Protobuf(e)
+    }
+}
+
+impl<'a> LabelEncoder<'a> {
+    pub fn encode_label_key(&mut self) -> LabelKeyEncoder {
+        match self {
+            LabelEncoder::Text(e) => e.encode_label_key().into(),
+            LabelEncoder::Protobuf(e) => e.encode_label_key().into(),
+        }
+    }
+}
+
+pub trait EncodeLabelKey {
+    fn encode(&self, encoder: &mut LabelKeyEncoder) -> Result<(), std::fmt::Error>;
+}
+
+pub enum LabelKeyEncoder<'a> {
+    Text(text::LabelKeyEncoder<'a>),
+    Protobuf(proto::LabelKeyEncoder<'a>),
+}
+
+impl<'a> From<text::LabelKeyEncoder<'a>> for LabelKeyEncoder<'a> {
+    fn from(e: text::LabelKeyEncoder<'a>) -> Self {
+        Self::Text(e)
+    }
+}
+
+impl<'a> From<proto::LabelKeyEncoder<'a>> for LabelKeyEncoder<'a> {
+    fn from(e: proto::LabelKeyEncoder<'a>) -> Self {
+        Self::Protobuf(e)
+    }
+}
+
+impl<'a> std::fmt::Write for LabelKeyEncoder<'a> {
+    fn write_str(&mut self, s: &str) -> std::fmt::Result {
+        for_both!(self, e, e.write_str(s))
+    }
+}
+
+impl<'a> LabelKeyEncoder<'a> {
+    pub fn encode_label_value(self) -> Result<LabelValueEncoder<'a>, std::fmt::Error> {
+        for_both_map!(self, e, e.encode_label_value(), LabelValueEncoder)
+    }
+}
+
+pub trait EncodeLabelValue {
+    fn encode(&self, encoder: &mut LabelValueEncoder) -> Result<(), std::fmt::Error>;
+}
+
+pub enum LabelValueEncoder<'a> {
+    Text(text::LabelValueEncoder<'a>),
+    Protobuf(proto::LabelValueEncoder<'a>),
+}
+
+impl<'a> LabelValueEncoder<'a> {
+    #[must_use]
+    fn finish(self) -> Result<(), std::fmt::Error> {
+        for_both!(self, e, e.finish())
+    }
+}
+
+impl<'a> std::fmt::Write for LabelValueEncoder<'a> {
+    fn write_str(&mut self, s: &str) -> std::fmt::Result {
+        for_both!(self, e, e.write_str(s))
+    }
+}
+
+impl<T: EncodeLabel> EncodeLabelSet for &[T] {
+    fn encode(&self, mut encoder: LabelSetEncoder) -> Result<(), std::fmt::Error> {
+        if self.is_empty() {
+            return Ok(());
+        }
+
+        for label in self.iter() {
+            label.encode(encoder.encode_label())?
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: EncodeLabel> EncodeLabelSet for Vec<T> {
+    fn encode(&self, encoder: LabelSetEncoder) -> Result<(), std::fmt::Error> {
+        self.as_slice().encode(encoder)
+    }
+}
+
+impl EncodeLabelSet for () {
+    fn encode(&self, _encoder: LabelSetEncoder) -> Result<(), std::fmt::Error> {
+        Ok(())
+    }
+}
+
+impl<K: EncodeLabelKey, V: EncodeLabelValue> EncodeLabel for (K, V) {
+    fn encode(&self, mut encoder: LabelEncoder) -> Result<(), std::fmt::Error> {
+        let (key, value) = self;
+
+        let mut label_key_encoder = encoder.encode_label_key();
+        key.encode(&mut label_key_encoder)?;
```

Comment:
In serde, I believe this is flipped: The serializer has functions for writing certain types, like `Serializer::serialize_str`: https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_str

Would it make sense to do the same thing here? Our "data format" are metrics so I think it would make sense if an encoder has a function "write_label" and you can pass a string to it.

I still need to dig into this in detail so excuse if it doesn't make any sense!